src/fuselage/CCPACSPressureBulkheadAssemblyPosition.cpp

```diff
@@ -16,18 +16,152 @@
 
 #include "CCPACSPressureBulkheadAssemblyPosition.h"
 
+#include <gp_Pln.hxx>
+#include <BRepBuilderAPI_MakeEdge.hxx>
+#include <BRepBuilderAPI_MakeFace.hxx>
+#include <BRepBuilderAPI_MakeWire.hxx>
+#include <TopExp.hxx>
+#include <Bnd_Box.hxx>
+#include <BRepBndLib.hxx>
+#include <gp_Ax3.hxx>
+#include <TopTools_IndexedMapOfShape.hxx>
+
+#include "CCPACSFuselage.h"
+#include "CCPACSFrame.h"
+#include "CCPACSPressureBulkhead.h"
+#include "CCPACSFuselageStringerFramePosition.h"
+#include "CTiglUIDManager.h"
+#include "tiglcommonfunctions.h"
+
 namespace tigl
 {
-
 CCPACSPressureBulkheadAssemblyPosition::CCPACSPressureBulkheadAssemblyPosition(CCPACSPressureBulkheadAssembly* parent,
                                                                                CTiglUIDManager* uidMgr)
     : generated::CPACSPressureBulkheadAssemblyPosition(parent, uidMgr)
 {
 }
 
+void CCPACSPressureBulkheadAssemblyPosition::SetFrameUID(const std::string& value)
+{
+    generated::CPACSPressureBulkheadAssemblyPosition::SetFrameUID(value);
+    Invalidate();
+}
+
+void CCPACSPressureBulkheadAssemblyPosition::SetPressureBulkheadElementUID(const std::string& value)
+{
+    generated::CPACSPressureBulkheadAssemblyPosition::SetPressureBulkheadElementUID(value);
+    Invalidate();
+}
+
 void CCPACSPressureBulkheadAssemblyPosition::Invalidate()
 {
-    invalidated = true;
+    m_geometry      = boost::none;
 }
 
+TopoDS_Shape CCPACSPressureBulkheadAssemblyPosition::GetGeometry()
+{
+    if (!m_geometry)
+        BuildGeometry();
+    return m_geometry.value();
+}
+
+void CCPACSPressureBulkheadAssemblyPosition::BuildGeometry()
+{
+    CCPACSFuselage& fuselage = *m_parent->GetParent()->GetParent();
+    CCPACSFrame& frame = m_uidMgr->ResolveObject<CCPACSFrame>(m_frameUID);
+
+    if (frame.GetFramePositions().size() == 1) {
+        // one frame position
+
+        TopTools_IndexedMapOfShape edgeMap;
+        TopExp::MapShapes(frame.GetGeometry(true), TopAbs_EDGE, edgeMap);
+
+        if (edgeMap.Extent() != 1) {
+            throw CTiglError("1D frame geometry should have exactly one edge");
+        }
+
+        TopoDS_Edge edge = TopoDS::Edge(edgeMap(1));
+        TopoDS_Wire wire = BRepBuilderAPI_MakeWire(edge); //turn the edge into wire
+        TopoDS_Face Face = BRepBuilderAPI_MakeFace(wire); //make the face from the wire
+        m_geometry       = Face;
+    } else if (frame.GetFramePositions().size() >= 2) {
+        // two ore more frame positions
```

Comment:
I don't quite understand, how the geometry is generated. In case there are different frame positions, the frame itself is not coplanar anymore, right? 
Then you use some center point and somehow connect it the the frame wire to make a face. How does this work? Do you create many triangles?

Reply:
The pressure bulkhead definition relies on the frame definition.
The frame has two cases:
 - one Position -> full frame
 - two or more positions -> Frame between the positions

This behaviour has to be reflected by the pressure bulkhead
 - one Position of the reference Frame -> full bulkhead
 - two or more positions -> The bulkhead is build up from triangles between the FuselageStringerFramePosition reference points (defined by Xref/Yref/Zref) and the  FuselageStringerFramePosition fuselage intersection points (defined by the angle)

Small example:


[PB3Pos.zip](https://github.com/DLR-SC/tigl/files/1924784/PB3Pos.zip)

Reply:
i have to apologize as it looks there is still a final cleanup missing.